feat:  RabbitMQ에서 DB로 데이터 저장

gnoesnooj opened the pull request:
rabbitmq 에서 메세지 컨슘 후 elasticsearch 와 postgre DB에 각각 저장해준다.

- 클래스나 변수명 네이밍이 고민 되었습니다
- 패키징이 올바르게 되었는지 확신이 안 들었습니다.

0607 수정 PR 완료 하였습니다.

gnoesnooj committed:
chore: key 관리 파일 gitignore 에 추가

gnoesnooj committed:
feat: Rabbitmq 에서 컨슘 후 DB 저장 구현

 - Postgre 저장
 - ElasticSearch 저장

joyfulviper commented on src/main/java/papyrus/article/infrastructure/rabbitmq/RabbitmqConsumer.java:
```diff
@@ -0,0 +1,37 @@
+        PostgreArticle pgArticle = message.toPostgreArticle();
+
+        postgreArticleRepository.save(pgArticle);
+        elasticSearchArticleRepository.save(esArticle);
```
컨슈머에서 바로 dao 계층에 접근하는건 좀 위험한 방법 같기도 하고 확장에도 닫혀있는 코드일 것 같습니다. 캡슐화에도 문제가 있어 보이고요

또한 트랜잭션 처리도 컨슈머에서 하는건 컨슈머가 너무 많은 역할을 한다고 생각하는데 컨슈머 클래스는 서비스 계층을 이용하게 하는게 어떨까요?? 마치 프레젠테이션 계층에서 서비스 계층을 이용하는 것 처럼요

추후 사용자에게 뉴스 기사를 제공하는 로직도 만들어야 할텐데 컨트롤러에서 바로 dao 계층을 호출하는 것은 너무 위험할 것 같아요

제 생각에는 아래처럼 작성하고 컨슈머 클래스, 컨트롤러 클래스에서 이 서비스 클래스를 이용하는게 더 나을 것 같은데 어떨까요??

+@Bean과 같은 설정 내용은 따로 설정하는 클래스로 빼주는게 좋을 것 같아요 나중에 Rabbitmq에 대해서 예외 설정이나 커스텀 해야하는 설정을 등록해야 할수도 있으니..  클래스를 만들기 전에 해당 클래스의 역할과 책임을 어디까지 경계지어야 할지 명확하게 결론짓고 작성해봐도 괜찮을거 같아요..!!
```
@Service
@RequiredArgsConstructor
public class ArticleService {
    private final PostgreArticleRepository postgreArticleRepository;
    private final ElasticSearchArticleRepository elasticSearchArticleRepository;

    @Transactional
    public void save(Message message) {
        postgreArticleRepository.save(message.toPostgreArticle());
        elasticSearchArticleRepository.save(message.toElasticSearchArticle());
    }
}
```

gnoesnooj replied:
ArticleService 라는 클래스를 만들어서 추후에 제공할 기능을 담당하도록 생각하신거에 동의합니다. 확장에 대해서 더 생각했어야 했는데 감사합니다 ! 또한 bean 설정 역시 config 파일 생성 후 담당하도록 하는 것이 좋을 것 같아요 !

gnoesnooj replied:
Consumer 클래스가 컨트롤러 계층에 해당된다고 생각하신 이유를 여쭤봐도 될까요 ? 

저는 service나 infrastructure 둘 중 어디에 해당될까 로만 생각했고, article 도메인 내에서는 service에는 해당되지 않는다고 생각해서 infrastructure 로 판단하고 패키징을 해줬습니다. 제가 뭘 놓친게 있을까요 ?

joyfulviper replied:
프레젠테이션 계층에 대한 내용은 예시를 든 것 뿐입니다!

이 계층이 프레젠테이션일지 인프라스트럭처일지는 생각하는 관점에 따라 달라질 수 있다고 생각하는데 이 계층의 주된 역할은 외부로부터 데이터를 수신하고 필요한 처리를 위해 비즈니스 로직 계층(여기서는 ArticleService)에게 전달하는 것이여서 마치 프레젠테이션 계층과 유사한 역할을 했기 때문에 프레젠테이션 계층을 예로 들었습니다!

gnoesnooj replied:
전 presentation이 papyrus API를 사용하는 사용자들과 상호작용을 하는 계층이라고 생각했고, 내부적으로 API제공을 위해 메세지 큐를 통해 데이터를 DB에 저장하는 역할은 infrastructure에 어울릴 것 같다고 생각했습니다. 좀 더 생각해봐야 할 것 같네요 ! 새로운 관점에 대해 알게 됐어요. 감사합니다 !

graygreat replied:
어느 계층에 둘까에 대해서 고민중이시군요..
저는 application layer에 두면 어떨까 생각을 해봅니다.

Infrastructure layer을 보면 "데이터베이스, 외부 서비스, 메시지 큐 등과의 상호작용을 처리" 라고 나와있긴 해요.
그래서 인프라스트럭처 계층에 두면 되지 않을까 생각할 수 있는데, 저는 아래와 같은 이유로 application layer에 둬도 되지 않을까 생각을 해봅니다.

* 비즈니스 로직이 있지 않을까?
  ㄴ 메시지 큐를 소비하는 로직은 아무래도 비즈니스 로직과 밀접하게 연관되어 있을 것 같아요. 메시지 큐는 주로 상태 변화나 이벤트 기반 처리를 지원하니 application layer에 위치해서 비즈니스 로직과 결합된다고 생각이 듭니다.

* 도메인 처리
  ㄴ 메시지 큐에 담기는 메시지는 도메인 객체나 비즈니스 서비스에서 처리되어야 할 수 있을 것 같아요.

* 유연성
  ㄴ 메시지 큐를 consume하는 로직이 infrastructure layer에 위치하면, application layer와 강결합 될 수 있을 것 같아요.
이건 유연성 측면에서 좋지 않을 것 같아요. application layer에 메지시 큐 consume 로직을 두면, infrastructure layer을 추상화해서 유연한 구조를 유지할 수 있지 않을까요?

joyfulviper commented on src/main/java/papyrus/article/infrastructure/elasticsearch/config/ElasticSearchConfig.java:
```diff
@@ -0,0 +1,24 @@
+    @Override
+    public RestHighLevelClient elasticsearchClient() {
+        ClientConfiguration clientConfiguration = ClientConfiguration.builder()
+                .connectedTo("localhost:9200")
```
여기 localhost:9200 부분은 yml파일에서 받는게 나을 것 같아요 만약 배포하게 된다면 이 부분 코드를 수정해야 할수도 있으니

joyfulviper commented on src/main/java/papyrus/article/domain/PostgreArticle.java:
```diff
@@ -0,0 +1,33 @@
+    private String press;
+
+    @Column(columnDefinition = "text")
+    private List<String> paragraphs;
```
비록 메시지 큐에 저장될 때 크롤링 과정에서 메시지가 List로 저장되더라도 필드는 List가 아닌 String으로 저장하는게 나을 것 같아요.

 왜냐하면 여기서 paragraphs를 gpt에게 보내주게 될텐데 List에 있는거를 다 꺼내서 붙여주는 것 보다는 애초에 저장을 String 형태로 저장하는게 더 깔끔할 것 같습니다. 또한 형태소 분석기를 사용할텐데 List로 나눠서 저장하하면 처리하기 힘들 것 같고요

아래 메시지를 entity로 변환하는 코드중 List 저장 부분을 Stream API를 이용해서 String으로 변환해서 저장할 수 있을 것 같은데 어떨까요??
리스트를 필드로 같는것 보다는 String을 필드로 같는게 db 필드상으로도 깔끔해질 것 같습니다.

gnoesnooj replied:
여기서도 확장과 추후 기능 적용에 대해 깊이 생각하지 못한 티가 나네요 ㅠㅠ stream 을 통해서 변환하는 것을 한번 적용해보겠습니다 !

graygreat commented on src/main/java/papyrus/article/domain/ElasticSearchArticle.java:
```diff
@@ -0,0 +1,28 @@
+@AllArgsConstructor
+@Builder
+@Document(indexName = "articles")
+public class ElasticSearchArticle {
```
클래스명을 다른걸로 바꿔주세요.
다른 프로젝트들에서 어떻게 사용하는지 한번 찾아보시죠

graygreat commented on src/main/java/papyrus/article/domain/ElasticSearchArticle.java:
```diff
@@ -0,0 +1,28 @@
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
+@Document(indexName = "articles")
```
@Document는 무엇을 의미하나요?

gnoesnooj replied:
ElasticSearch 에서 사용하는 어노테이션입니다. jpa가 엔티티 클래스를 @entity 를 통해서 매핑해주는 것처럼, elasticseach에서 문서를 매핑해주기 위해 사용하는 어노테이션으로 알고 있습니다.

graygreat commented on src/main/java/papyrus/article/domain/PostgreArticle.java:
```diff
@@ -0,0 +1,33 @@
+@AllArgsConstructor
+@NoArgsConstructor
+@Builder
+public class PostgreArticle {
```
이것도 이름 바꿔주세요.

추가로 공통되는 필드 많은데 어떻게 해결할 수 있을까요?

gnoesnooj replied:
추상 클래스와 상속을 통해 적용해보는걸 시도해보겠습니다.

graygreat commented on src/main/java/papyrus/article/domain/repository/PostgreArticleRepository.java:
```diff
@@ -0,0 +1,7 @@
+import org.springframework.data.jpa.repository.JpaRepository;
+import papyrus.article.domain.PostgreArticle;
+
+public interface PostgreArticleRepository extends JpaRepository<PostgreArticle, Long> {
```
postgre 말고 다른 db를 사용하게 되면 이 클래스 이름 바꿔야할까요?

gnoesnooj replied:
db 교체가 필요할 시, 어댑터 패턴을 통해서 해결할 수 있을 것 같습니다.

graygreat commented on src/main/java/papyrus/article/infrastructure/rabbitmq/dto/RabbitmqMessage.java:
```diff
@@ -0,0 +1,44 @@
+@NoArgsConstructor
+@Builder
+@Getter
+public class RabbitmqMessage {
```
rabbitmq에서 es랑 postgre를 의존하고 있는데 좋은 구조일까요?

gnoesnooj replied:
구조 변경이나 repository 변경이 있을 시 코드 수정이 많이 일어날 것 같습니다..

entity를 변환해주는 enum과 같은 팩토리 클래스를 적용해보면 좋을 것 같습니다.

gnoesnooj committed:
refactor : PR #11 수정 사항에 대한 리팩토링

- 기존 DB이름을 포함하던 Entity의 Renaming
- RabbitmqMessage 의 entity 생성 전략 변경 - 팩토리 메소드 패턴 적용
- Repository의 추상화 - 어댑터 패턴 적용
- RabbitmqConsumer의 계층 변경 - infrasturcture -> application

joyfulviper commented:
전보다 훨씬 나아진거 같습니다 굳굳
